chore: throw semantic error for invalid arguments to intrinsic `present`

YBansal95 opened the pull request:
fixes #6074

```
program main
  call a()
  contains
  subroutine a(ii)
    integer ,optional :: ii
    integer :: mm
    if(present(mm)) print *, "bug"
  end subroutine a

end program main
```

Output:
```
semantic error: Argument to 'present' must be an optional dummy argument
 --> temp.f90:7:8
  |
7 |     if(present(mm)) print *, "bug"
  |        ^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

kmr-srbh reviewed (changes requested):
Thanks for the PR @YBansal95!

I request you to do the following:
- Please add text `fixes #6074` in the PR description to link this PR with the issue. This will ensure the issue gets closed automatically when this PR is merged. You may also update the PR title to something like "chore: throw semantic error for invalid arguments to intrinsic `present`". :)
- As we are adding a semantic error here, please add a test case which throws this error inside `tests/errors/continue_compilation_1.f90` and run `./run_tests.py -u` to update the test reference. Please push the created reference  file.

kmr-srbh commented on tests/reference/ast-parse_without_program_line-0b3170b.stderr:
```diff
@@ -0,0 +1,5 @@
+syntax error: Statement or Declaration expected inside program, found Expression
```
I think you are using an older built version of `lfortran`. I request you to do `git clean -dfx` and do a clean build with the latest `main` to generate the references. Then please delete this file.

YBansal95 replied:
Oh ok

YBansal95 replied:
I've fetched the latest changes and resolved conflicts.

kmr-srbh replied:
Please delete this file. I think we should be done. :)

YBansal95 replied:
Done. It took me some time to figure out git pull, rebase and stuff 😅 Thanks for the patience.

assem2002 commented:
Thanks @YBansal95 . The changes are great and covers more errors. You just resolve the conflicts and handle the test issues that Saurabh mentioned, and it'll be ready to merge.

YBansal95 commented on src/libasr/pass/intrinsic_functions.h:
```diff
@@ -1795,19 +1795,76 @@ namespace Present {
+            append_error(diag, 
+                "Argument to 'present' must be an optional dummy argument", 
+                loc);
+            return nullptr;
```
Is it ok now?

YBansal95 replied:
Which 5 errors? There are 4 error checks that I have added in `verify_args` and `create_Present` functions.

kmr-srbh commented:
@YBansal95 I think the changes in the branch overwrite the existing references after conflict resolution. Let me give you a a hand with this. :)

kmr-srbh commented:
@YBansal95 I updated the commit messages and rebased the branch on top of latest `main` after resolving conflicts. Please rebase your local branch using `git pull --rebase` for further development.

YBansal95 commented on tests/reference/asr-continue_compilation_1-04b6d40.stderr:
```diff
@@ -286,3 +286,27 @@ semantic error: string expression, expected logical
+   --> tests/errors/continue_compilation_1.f90:139:14
+    |
+139 |     print *, present(a)
+    |              ^^^^^^^^^^ 
```
Ok, I'll look into all these tests.

YBansal95 replied:
Can you elaborate a bit on how to split these errors?

YBansal95 commented:
> @YBansal95 I updated the commit messages and rebased the branch on top of latest `main` after resolving conflicts. Please rebase your local branch using `git pull --rebase` for further development.

Thanks a lot @kmr-srbh

YBansal95 commented on src/libasr/pass/intrinsic_functions.h:
```diff
@@ -1795,19 +1795,76 @@ namespace Present {
     static inline void verify_args(const ASR::IntrinsicElementalFunction_t& x, diag::Diagnostics& diagnostics) {
         if (x.n_args != 1)  {
-            ASRUtils::require_impl(false, "Unexpected number of args, Present takes 1 arguments, found " + std::to_string(x.n_args), x.base.base.loc, diagnostics);
+            ASRUtils::require_impl(false, "Unexpected number of args, Present takes 1 argument, found " + std::to_string(x.n_args), x.base.base.loc, diagnostics);
```
I added test case for this `print *, present(a, b)`. But I'm not sure from where it's invoking the error message `semantic error: No matching signature found for intrinsic present`.
